Added --chmod for ADD and COPY commands. Fixes #2850 and #1751

mschneider82 opened the pull request:
Fixes #2850 and #1751

**Description**

I have added --chmod feature for ADD and COPY command.

```
[36mINFO[0m[0004] ADD --chmod=644 https://xxx/test.gpg /etc/apt/trusted.gpg.d/ 
[36mINFO[0m[0004] RUN ls -la /etc/apt/trusted.gpg.d/test.gpg 
[36mINFO[0m[0007] Cmd: /bin/sh                                 
[36mINFO[0m[0007] Args: [-c ls -la /etc/apt/trusted.gpg.d/test.gpg] 
[36mINFO[0m[0007] Running: [/bin/sh -c ls -la /etc/apt/trusted.gpg.d/test.gpg] 
-rw-r--r-- 1 root root 1545 May  5  2023 /etc/apt/trusted.gpg.d/test.gpg
```

**Submitter Checklist**

These are the criteria that every PR should meet, please check them off as you
review them:

- [X] Includes [unit tests](../DEVELOPMENT.md#creating-a-pr)
- [X] Adds integration tests if needed.

**Reviewer Notes**

- [ ] The code flow looks good. 
- [ ] Unit tests and or integration tests added.


**Release Notes**

```
- kaniko now supports setting --chmod on ADD and COPY commands.
```

mschneider82 commented:
@aaron-prindle can you please review?

aaron-prindle commented:
@mschneider82, the PR is LGTM.  I'm seeing some test failures but I believe they are flakes and unrelated.  Can you rebase the PR and resubmit to see if we can get the CI/CD tests green?  Thanks

mschneider82 commented:
Still fails but i have tested the new Integration with LOCAL env set.

Error below from Integration Test on GitHub, the waiting is strange, are there some time Limits?

`            INFO[0000] Running: [/bin/sh -c test "$(stat -c "%a" /path/file666)" = "666"] 
            INFO[0000] Taking snapshot of full filesystem...        
            INFO[0000] No files were changed, appending empty layer to config. No layer added to image. 
            INFO[0000] RUN test "$(stat -c "%a" /path/file)" = "775" 
            INFO[0000] Cmd: /bin/sh                                 
            INFO[0000] Args: [-c test "$(stat -c "%a" /path/file)" = "775"] 
            INFO[0000] Running: [/bin/sh -c test "$(stat -c "%a" /path/file)" = "775"] 
            error building image: error building stage: failed to execute command: waiting for process to exit: exit status 1
FAIL
FAIL	github.com/GoogleContainerTools/kaniko/integration	488.942s
FAIL`

mschneider82 commented:
I have removed the mentioned checks for default behaiviour for the ADD and COPY without a --chmod command. the integration test now only checks the --chmod feature.

mschneider82 commented:
I dont see whats wrong with the Integration test

aaron-prindle commented:
I believe this is a test flake in our integration tests @mschneider82.  In the past 2-3 days I have seen this:
```
        integration_test.go:609: Couldn't get details from image reference for (localhost:5000/docker-dockerfile_test_copyadd_chmod): Couldn't get reference to image localhost:5000/docker-dockerfile_test_copyadd_chmod from daemon: Error response from daemon: No such image: localhost:5000/docker-dockerfile_test_copyadd_chmod:latest
```

I am still trying to root cause this but currently I am not sure what is causing this.  If I can identify the issue soon I will submit a fix and rebase and if not I think it makes sense to merge as is as I don't believe the changes in this PR are related to the test failures

flake tracking issue:
https://github.com/GoogleContainerTools/kaniko/issues/3124

previous runs with this flake (the same PR would then pass later):
https://github.com/GoogleContainerTools/kaniko/actions/runs/8682504136/job/23807123764#step:7:73

mschneider82 commented:
Ok feel free to merge then, i use this patched Version since days without an issue

aaron-prindle commented:
Just merged, #3131.  If you rebase and push we can hopefully see some  additional error information

mschneider82 commented:
Ok it fails because the docker command does not support chmod without buildkit:

```
            the --chmod option requires BuildKit. Refer to https://docs.docker.com/go/buildkit/ to learn how to build images with BuildKit enabled
```

mschneider82 commented:
ok doing buildkit=1 for integration is a longer task to fix, but this should be a seperated task. The lagacy build will be deprecated, to it should compare to buildkit instead of legacy docker build.

mschneider82 commented:
I found the chmod container build, with docker it had 11 layers with kaniko just 5?

https://github.com/GoogleContainerTools/kaniko/actions/runs/8780308512/job/24089955636?pr=3119#step:7:765

To debug this it would be helpful if the build from kaniko would be also printed. currently in [images.go ](https://github.com/mschneider82/kaniko/blob/main/integration/images.go#L514C33-L514C35) the output is just printed when the image build fails, but not returned, in this case it would be helpful to return it and also print it when the layercheck has a difference detected.

mschneider82 commented:
@aaron-prindle OK i got it green, i had to enable DOCKER_BUILDKIT=1 for just this single integration test via `envsMap` in `images.go` also added the layers difference of 6.  I have rebased everything. In future i think many more tests need to be fixed to use buildkit instead of classic build.

aaron-prindle reviewed (approved):
Thanks for your detailed explanations and investigation here @mschneider82.  Will update #3124 with your findings.

LGTM, merging
